src/cryptography/hazmat/primitives/serialization/ssh.py

```diff
@@ -616,14 +656,32 @@ def load_ssh_private_key(
             )
         if kdfname != _BCRYPT:
             raise UnsupportedAlgorithm(f"Unsupported KDF: {kdfname!r}")
-        blklen = _SSH_CIPHERS[ciphername_bytes][3]
+        blklen = _SSH_CIPHERS[ciphername_bytes].block_len
+        tag_len = _SSH_CIPHERS[ciphername_bytes].tag_len
+        # load secret data
+        edata, data = _get_sshstr(data)
+        # see https://bugzilla.mindrot.org/show_bug.cgi?id=3553 for
+        # information about how OpenSSH handles AEAD tags
+        if _SSH_CIPHERS[ciphername_bytes].is_aead:
+            tag = bytes(data)
+            if len(data) != tag_len:
+                raise ValueError("Corrupt data: invalid tag length for cipher")
+        else:
+            _check_empty(data)
         _check_block_size(edata, blklen)
         salt, kbuf = _get_sshstr(kdfoptions)
         rounds, kbuf = _get_u32(kbuf)
         _check_empty(kbuf)
         ciph = _init_cipher(ciphername_bytes, password, salt.tobytes(), rounds)
-        edata = memoryview(ciph.decryptor().update(edata))
+        dec = ciph.decryptor()
+        edata = memoryview(dec.update(edata))
+        if _SSH_CIPHERS[ciphername_bytes].is_aead:
+            assert isinstance(dec, AEADDecryptionContext)
+            _check_empty(dec.finalize_with_tag(tag))
```

Comment:
We really should have been calling finalize in the other case. Can you add an else that calls it and does `edata += dec.finalize()`?